src/test/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcherSpec.scala

```diff
@@ -124,6 +129,95 @@ class BlockFetcherSpec extends TestKit(ActorSystem("BlockFetcherSpec_System")) w
       peersClient.expectMsgClass(classOf[BlacklistPeer])
       peersClient.expectMsgPF() { case PeersClient.Request(msg, _, _) if msg == firstGetBlockHeadersRequest => () }
     }
+
+    "should not append new blocks if the received data does not match" in new TestSetup {
+
+      // Important: Here we are forcing the mismatch between request headers and received bodies
+      override lazy val validators = new MockValidatorsFailingOnBlockBodies
+
+      startFetcher()
+
+      val getBlockHeadersRequest =
+        GetBlockHeaders(Left(1), syncConfig.blockHeadersPerRequest, skip = 0, reverse = false)
+      peersClient.expectMsgPF() { case PeersClient.Request(msg, _, _) if msg == getBlockHeadersRequest => () }
+
+      val chain = BlockHelpers.generateChain(syncConfig.blockHeadersPerRequest, FixtureBlocks.Genesis.block)
+      val getBlockHeadersResponse = BlockHeaders(chain.map(_.header))
+      peersClient.reply(PeersClient.Response(fakePeer, getBlockHeadersResponse))
+
+      val getBlockBodiesRequest = GetBlockBodies(chain.map(_.hash))
+      peersClient.expectMsgPF() { case PeersClient.Request(msg, _, _) if msg == getBlockBodiesRequest => () }
+
+      // This response will be invalid given we are using a special validator!
+      val getBlockBodiesResponse = BlockBodies(chain.map(_.body))
+      peersClient.reply(PeersClient.Response(fakePeer, getBlockBodiesResponse))
+
+      peersClient.expectMsgClass(classOf[BlacklistPeer])
+      peersClient.expectMsgPF() { case PeersClient.Request(msg, _, _) if msg == getBlockBodiesRequest => () }
```

Comment:
what about test like:
  - fetch one batch of blocks
  - for the second batch send different bodies (so even proper validator fails)
  - pick blocks and verify that only these from first batch are responded

Basically same issue with these tests - lack of explicit assertions or at least comments, why it's enough to check for certain message/lack of it.

Reply:
The main issue i see with the test you propose is that i would need the real block validator, and that make things more complex. So i will opt for improve current test assertions if you are ok with that.

Reply:
(main issue is current impl of `bodiesAreOrderedSubsetOfRequested`, given i can't use a simple mock validator to match headers and bodies)